### PR TITLE
Flatten artifacts on load

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This is a draft solution to #62, #45, #20 & #78.

Results are flattened on upload into three tables:
- one for run result files (same granularity as artifacts)
- one for the flattened results together
- one for the flattened manifest elements (with sources, exposures compressed in with other nodes)

This upload is facilitated by a new `V2` upload macro, and then the results of both this approach and the old `V1` upload method are unioned together to give teams flexibility on which they use.

This should bypass field size limits (because we use `insert into`), and preloads the compute load of flattening large JSON blobs so allow larger projects to work well.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
...
+1. To initially create these tables, execute `dbt run-operation create_artifact_resources`
+   ([source](macros/create_artifact_resources.sql)). This will create a stage and a set of tables in
+   the `{{ target.database }}.dbt_artifacts` schema — you can override the database, schema and table
+   names using the variables listed in the Installation section, above.
+
+2. Add [operations](https://docs.getdbt.com/docs/building-a-dbt-project/hooks-operations/#operations)
+   to your production run to load files into your table.
+   
+   **V2 Macro**: Use the `upload_artifacts_v2` macro ([source](macros/upload_artifacts.sql)). You only
+   need to run the macro after `run`, `test`, `seed` or `build` operations.
```
Comment (NiallRees) [inline README.md]:
```suggestion
   need to run the macro after `build`, `snapshot`, `run`, `test`, `seed` or `build` operations.
```

~~The new approach means we just support the build command automatically right?~~ Can we just remove this filter now that we're filtering by resource type in each model? https://github.com/brooklyn-data/dbt_artifacts/blob/main/models/staging/stg_dbt__run_results.sql#L20

File: macros/create_artifact_resources.sql [hunk 2]
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
 file_format = (type = json);
 {% endset %}
 
+{% set create_new_stage_query %}
+create stage if not exists {{ artifact_stage }}
+file_format = ( type =  json );
```
Comment (NiallRees) [inline macros/create_artifact_resources.sql]:
```suggestion
file_format = (type = json);
```

File: macros/create_artifact_resources.sql [hunk 3]
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
...
     artifact_type string
 );
+{% endset %}
+
+{% set create_results_query %}
+create table if not exists {{ src_results }} (
+    command_invocation_id string,
+    dbt_cloud_run_id int,
+    artifact_run_id string,
+    artifact_generated_at timestamp_ntz,
```
Comment (NiallRees) [inline macros/create_artifact_resources.sql]: Looks like all dbt artifact timestamps are in UTC: https://github.com/brooklyn-data/dbt_artifacts/blob/main/integration_test_project/artifacts/schema_v4/build/run_results.json#L5
```suggestion
    artifact_generated_at timestamp_tz,
```

File: macros/create_artifact_resources.sql [hunk 4]
```diff
@@ -1,6 +1,13 @@
 {% macro create_artifact_resources() %}
 
 {% set src_dbt_artifacts = source('dbt_artifacts', 'artifacts') %}
+{% set artifact_stage = var('dbt_artifacts_stage', 'dbt_artifacts_stage') %}
+
+{% set src_results = source('dbt_artifacts', 'dbt_run_results') %}
+{% set src_result_nodes = source('dbt_artifacts', 'dbt_run_result_nodes') %}
```
Comment (NiallRees) [inline macros/create_artifact_resources.sql]:
```suggestion
{% set src_run_results = source('dbt_artifacts', 'dbt_run_results') %}
{% set src_run_results_nodes = source('dbt_artifacts', 'dbt_run_results_nodes') %}
```

Just to keep the names aligned with the artifacts exactly

File: macros/create_artifact_resources.sql [hunk 5]
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
...
+    package_name string,
+    node_json variant
+);
+{% endset %}
 
-{% do log("Creating Stage: " ~ create_stage_query, info=True) %}
+{% do log("Creating Old Stage: " ~ create_stage_query, info=True) %}
 {% do run_query(create_stage_query) %}
 
+{% do log("Creating New Stage: " ~ create_new_stage_query, info=True) %}
```
Comment (NiallRees) [inline macros/create_artifact_resources.sql]: Let's version these like you've done elsewhere

File: macros/upload_artifacts_v2.sql [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                run_results.$1:metadata as metadata,
+                run_results.$1:args as args,
+                run_results.$1:elapsed_time::float as elapsed_time
+            from  @{{ artifact_stage }} as run_results
+
+        )
+        
+        select
+            metadata:invocation_id::string as command_invocation_id,
+            metadata:env:DBT_CLOUD_RUN_ID::int as dbt_cloud_run_id,
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql:32]:
```suggestion
            metadata:env:dbt_cloud_run_id::int as dbt_cloud_run_id,
```

File: macros/upload_artifacts_v2.sql [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+{% set results_query %}
+    begin;
+    insert into {{ src_results }}
+        with raw_data as (
+
+            select
+                run_results.$1:metadata as metadata,
+                run_results.$1:args as args,
+                run_results.$1:elapsed_time::float as elapsed_time
+            from  @{{ artifact_stage }} as run_results
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql]:
```suggestion
            from @{{ artifact_stage }} as run_results
```

File: macros/upload_artifacts_v2.sql [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            result.value:adapter_response:rows_affected::int as rows_affected,
+            -- Future proof by also storing the whole json body to use later if we need to.
+            result.value as result_json
+        from raw_data,
+            lateral flatten(input => results) as result;
+    commit;
+{% endset %}
+
+{% set manifest_nodes_query %}
+    begin;
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql]:
```suggestion
    begin;
```

Not sure that the begin; and commit; are necessary - the run_query macro handles all that to my knowledge

File: models/staging/sources.yml [hunk 9]
```diff
@@ -18,3 +19,89 @@ sources:
...
+            description: Timestamp of when the source artifact was generated.
+          - name: node_id
+            description: Unique id for the node, in the form of exposure.[package_name].[exposure_name]
+          - name: node_json
+            description: Raw JSON structure of the manifest node.
+      - name: dbt_run_manifest_exposures
+        identifier: "{{ var('dbt_artifacts_manifest_exposures_table', 'dbt_run_manifest_exposures') }}"
+        description: |
+          The source table containing the loaded and flattened exposures from manifest.json loaded artifacts. This belongs
+          to the _new style_ upload. See the README for more info.
```
Comment (NiallRees) [inline models/staging/sources.yml]: Let's avoid relative versioning and go for v1 and v2 like elsewhere

File: macros/upload_artifacts_v2.sql (hunk 6 above)
Comment (alanmcruickshank) [reply macros/upload_artifacts_v2.sql:32]: I think this is actually case sensitive - might be remembering wrong but I'll add a comment to that effect.

File: README.md (hunk 1 above)
Comment (alanmcruickshank) [reply README.md]: Yep.... and you put `build` in the list twice 🤣

File: models/staging/sources.yml [hunk 10]
```diff
@@ -18,3 +19,62 @@ sources:
...
+            description: The id of the command which resulted in the source artifact's generation.
+          - name: artifact_run_id
+            description: A constructed ID to serve as a reliable identifier for a single run.
+          - name: artifact_generated_at
+            description: Timestamp of when the source artifact was generated.
+          - name: node_id
+            description: Unique id for the node, in the form of exposure.[package_name].[exposure_name]
+          - name: node_json
+            description: Raw JSON structure of the manifest node.
+
```
Comment (NiallRees) [inline models/staging/sources.yml]:
```suggestion
```

**Comment by alanmcruickshank** [comment]
@NiallRees - Check out what I've tried with the integration suite.

- I've added an optional `replace` argument for the create macro so we can get fresh artifact tables without duplicating the code.
- I've made a separate integration suite for the V2 upload. In theory we should have the same result whether we use the V1 or V2 upload. I don't know if we have any data tests in place - but if we did then they should work in either case.

File: tox.ini [hunk 11]
```diff
@@ -103,6 +104,21 @@ commands =
...
+    dbt build -s dbt_artifacts --full-refresh
+    dbt build -s dbt_artifacts
+
+[testenv:integration_v2]
+changedir = integration_test_project
+commands =
+    dbt deps
+    # Setting the replace arg to tru so that we get a clean set of tabels on each run.
+    dbt run-operation create_artifact_resources --args '\{replace: true\}'
+    dbt run-operation upload_dbt_artifacts_v2
```
Comment (NiallRees) [inline tox.ini]: Can we test uploading all of the files in the `schema_v4` directory here like the above integration tests do?

File: macros/create_artifact_resources.sql [hunk 12]
```diff
@@ -1,29 +1,115 @@
...
+
+{% set create_v2_stage_query %}
+create {% if replace %} or replace {% endif %} stage if not exists {{ artifact_stage }}
 file_format = (type = json);
 {% endset %}
 
-{% set create_table_query %}
-create table if not exists {{ src_dbt_artifacts }} (
+{% set create_v1_table_query %}
+create {% if replace %} or replace {% endif %} table if not exists {{ src_dbt_artifacts }} (
```
Comment (NiallRees) [inline macros/create_artifact_resources.sql]: Unfortunately `or replace` and `if not exists` can't be used in the same statement: https://docs.snowflake.com/en/sql-reference/sql/create.html#general-usage-notes

Comment (NiallRees) [reply macros/create_artifact_resources.sql]: For the purposes of these automated integration tests the `replace` doesn't matter as each runs in a separate schema named by the commit hash. Appreciate it's useful for local testing though.

Comment (alanmcruickshank) [reply macros/create_artifact_resources.sql]: ahhhhh - understand. I thought it was _necessary_ for the CI tests! In that case I'm just going to revert the `replace` changes. I think it's almost a _dangerous_ feature if you don't need it.

File: tox.ini (hunk 11 above)
Comment (alanmcruickshank) [reply tox.ini]: 👍

**Comment by alanmcruickshank** [comment]
Ok I think this is ready for another review:
- I removed the `or replace` option entirely. I think it's a dangerous feature unless we need it for CI testing.
- I've added the `prefix` option into the new upload macro so we can test all the new artifacts.

I intentionally haven't added the `sources` file as one of the things we test - because none of the flatten code handles it. I don't know if this is an issue because I can't see any code in the rest of the package which actually _handles_ a sources file. I assume that would be tackled as a totally separate issue?

File: .github/workflows/ci_test_package.yml [hunk 13]
```diff
@@ -37,5 +37,8 @@ jobs:
       - name: Install tox
         run: python3 -m pip install tox
 
-      - name: Run Tests
+      - name: Run V1 Tests
         run: tox -e integration
+      
+      - name: Run V2 Tests
+        run: tox -e integration_v2
```
Comment (NiallRees) [inline .github/workflows/ci_test_package.yml]:
```suggestion
      - name: Run Tests
        run: tox -e integration
```

File: .github/workflows/main_test_models.yml [hunk 14]
```diff
@@ -33,5 +33,8 @@ jobs:
       - name: Install tox
         run: python3 -m pip install tox
 
-      - name: Run Tests
+      - name: Run V1 Tests
         run: tox -e integration
+      
+      - name: Run V2 Tests
+        run: tox -e integration_v2
```
Comment (NiallRees) [inline .github/workflows/main_test_models.yml]:
```suggestion
      - name: Run Tests
        run: tox -e integration
```
```

File: README.md [hunk 15]
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
 3. Run `dbt deps`.
 
 ## Generating the source table
```
Comment (NiallRees) [inline README.md]:
```suggestion
## Uploading the artifacts
```

File: README.md [hunk 16]
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
...
-This package requires that the source data exists in a table in Snowflake.
-
-### Option 1: Loading local files
+This package requires that the source data exists in Snowflake. There are two supported ways of doing this:
```
Comment (NiallRees) [inline README.md]:
```suggestion
This package uploads the artifact files into Snowflake. There are two supported ways of doing this:
```

File: README.md [hunk 17]
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
...
+- The _V2_ way of doing this which flattens the uploaded files on load. This supports files over
+  16MB (the limit of a variant field in snowflake) and also makes rebuilds of the materialised
+  models much faster because the JSON unpacking is done once on load. The downside of this approach
+  is that the upload is much heavier and more complex, as such we only directly support the
+  _"local file"_ method. Loading via cloud storage is also _possible_ but we recommend users
+  copy the method used in `upload_artifacts_v2.sql` to create their own approach.
+- The _V1_ or _legacy_ option, which uploads the files unprocessed. This affords much more flexibility
+  in their use, but the is subject to field size limits and higher compute loads to reprocess the
```
Comment (NiallRees) [inline README.md]:
```suggestion
  in their use, but is subject to field size limits and higher compute loads to reprocess the
```

**Review by NiallRees — APPROVED** [review]
Great job. Tests too!

**Pull request merged**